### ISSUE TITLE
FilePicker default directory

### ISSUE THREAD

**Issue by DevilDog2098** [dotnet/maui#9212]
### Description

I would love to have the ability to set a default directory in the FilePicker type.   i.e.  Set FilePicker to a specific directory when it displays to the user.

### Public API Changes

public class PickOptions
    {
        public PickOptions();

      
        public static PickOptions Default { get; }
    
    
        public static PickOptions Images { get; }
 
        public string? PickerTitle { get; set; }     

        public FilePickerFileType? FileTypes { get; set; }

       --> public string BaseDirectory? {get; set;}
    }
}

### Intended Use-Case

In my application I have the ability for users to Open and Backup/Restore project files.   Currently when "Open Project" is selected the FilePicker opens in the Documents directory.  Being able to set a default directory for the picker to open too would save a lot of support issues once a production release is pushed out.

**Comment by mavaddat** [comment]
I would like to see this feature as well. For my case, I have found that if the last-chosen path was one on the network, it can take a while to load the file picker if the user is not connected to the VPN. I would leverage a cancellation token to check that the file picker UI is shown within a reasonable time to set the starting path so it is the user's home (e.g., `%USERPROFILE%` on Windows or `$HOME` on Linux and macOS) if there is a timeout.

**Comment by taimo42** [comment]
I also would like to have this feature, so that the user doesn't need to take care about the directory anymore.

**Comment by mavaddat** [comment]
I have done some research, and it appears the problem with this feature request is that the platform-specific implementations of the file/folder picker are inconsistent about providing an option to specify the starting directory.

| Operating System | C# SDK Namespace for file picker | File/Folder picker allows specifying start directory | Qualififications/Caveats | API Documentation |
|------------------|----------------------------------|-----------------------------------------------------|--------------------------|-------------------|
| Windows          | `Windows.Storage.Pickers`        | No                                                  | The `SuggestedStartLocation` property only allows you to set predefined locations. | [Windows.Storage.Pickers](https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers) |
| Android          | `Android.Content`                | No                                                  | The `ACTION_GET_CONTENT` protocol does not support opening in a specific location. | [Android.Content](https://developer.android.com/reference/android/content/package-summary) |
| iOS              | `UIKit.UIDocumentPickerViewController` | Yes                                                 | You can specify the starting directory by setting the `directoryURL` property. However, this property has no effect in Mac apps built with Mac Catalyst. | [UIDocumentPickerViewController](https://developer.apple.com/documentation/uikit/uidocumentpickerviewcontroller) |
| Tizen            | `Tizen.Applications.FilePicker`  | No                                                  | The API does not provide a way to specify a custom start path for the file picker. | [Tizen.Applications.FilePicker](https://developer.tizen.org/dev-guide/csapi/api/Tizen.Applications.FilePicker.html) |
| MacCatalyst      | `UIKit.UIDocumentPickerViewController` | No                                                  | The `directoryURL` property has no effect in Mac apps built with Mac Catalyst. | [UIDocumentPickerViewController](https://developer.apple.com/documentation/uikit/uidocumentpickerviewcontroller) |
| macOS            | `AppKit.NSOpenPanel`             | No                                                  | The `NSOpenPanel` does not support opening in a specific location programmatically. | [NSOpenPanel](https://developer.apple.com/documentation/appkit/nsopenpanel) |
| tvOS             | `UIKit.UIDocumentPickerViewController` | Unknown                                             | I couldn't find specific information on how to specify a starting directory for the file picker in tvOS. | [UIDocumentPickerViewController](https://developer.apple.com/documentation/uikit/uidocumentpickerviewcontroller) |
| watchOS          | `WatchKit.WKInterfacePicker`     | No                                                  | In watchOS, you can use the Photos picker view to browse and select images and videos from the photo library. However, it doesn't seem like you can specify a starting directory. | [WKInterfacePicker](https://developer.apple.com/documentation/watchkit/wkinterfacepicker) |

Since the standard APIs provided by the OS C# SDK are inconsistent on this point, implementing this feature in a cross-platform UI framework like .NET MAUI would require developing custom file-picker libraries. I think going down this road would make MAUI _opinionated_ about UI implementations and be opposed to the generic, flexible design goals of .NET MAUI.

**Comment by DDHSchmidt** [comment]
@mavaddat I haven't checked yet all of the linked documentations, but isn't this somewhat covered in Android by the "EXTRA_INITIAL_URI" Parameter when calling the "[ACTION_OPEN_DOCUMENT](https://developer.android.com/training/data-storage/shared/documents-files#open-file)" Intent?

**Comment by mavaddat** [comment]
> Isn't this somewhat covered in Android by the "EXTRA_INITIAL_URI" Parameter when calling the "[ACTION_OPEN_DOCUMENT](https://developer.android.com/training/data-storage/shared/documents-files#open-file)" Intent?

Yes, it looks like you are correct. Although the ACTION_GET_CONTENT does not support opening in a specific location, [Intent#ACTION_OPEN_DOCUMENT](https://developer.android.com/reference/android/content/Intent#ACTION_OPEN_DOCUMENT) [Intent#ACTION_CREATE_DOCUMENT](https://developer.android.com/reference/android/content/Intent#ACTION_CREATE_DOCUMENT) [Intent#ACTION_OPEN_DOCUMENT_TREE](https://developer.android.com/reference/android/content/Intent#ACTION_OPEN_DOCUMENT_TREE) through the [EXTRA_INITIAL_URI](https://developer.android.com/reference/android/provider/DocumentsContract#EXTRA_INITIAL_URI) intent extra. I have updated the table above.

**Comment by EricHarmon** [comment]
I also need this feature. I just wrote my very first MAUI app (a short 300 lines of code) for a friend to assist him with a tedious, mundane task on his Mac. The very first thing he asked for when I sent him the application, was to be able to have it come up with a default folder when opening a file.

**Comment by mavaddat** [comment]
> To add more details to the excelent writeup [#9212 (comment)](https://github.com/dotnet/maui/issues/9212#issuecomment-1826125726):
> 
>     * Regarding the **Windows** platform, it seems one can actually work it around using this approach: [New File/Folder picker for WinAppSDK microsoft/WindowsAppSDK#88 (comment)](https://github.com/microsoft/WindowsAppSDK/issues/88#issuecomment-2604411467)

The problem is that the workaround (`IFileDialog::SetFolder` or `IFileDialog::SetDefaultFolder`) is not part of the standard `Windows.Storage.Pickers` API used in most .NET MAUI applications and requires using lower-level Windows APIs.

The `IFileDialog::SetFolder` API is a COM-based API native to C++ and is not directly available within .NET MAUI's C# abstractions or the `Windows.Storage.Pickers` namespace. Using this API in a .NET MAUI application would require significant effort and re-engineering. Here’s why:

### Challenges with Using `IFileDialog::SetFolder` in .NET MAUI

1. **Interop Layer**: 
   - The `IFileDialog::SetFolder` API is a native Windows COM API. To use it in .NET (and by extension, MAUI), you would need to create an interop layer using P/Invoke or COM interop to bridge between C++ and C#.
   - This requires understanding and managing COM objects, memory, and lifetimes, which adds complexity.

2. **Custom Implementation**:
   - MAUI is designed as a cross-platform framework. It abstracts platform-specific behavior into interfaces and renderers. 
   - To use `IFileDialog::SetFolder`, you would need to implement a **Windows-specific custom file picker interface**, bypassing the existing `Windows.Storage.Pickers`.

3. **Breaking Abstractions**:
   - Introducing a custom file picker implementation for Windows would deviate from MAUI’s philosophy of providing a unified, platform-agnostic API.
   - This could introduce inconsistencies and limit portability for developers expecting uniform behavior across platforms.

4. **Compatibility and Maintenance**:
   - Integrating this functionality could complicate future updates and maintenance. Any changes to the Windows API or MAUI framework could require additional work to keep the custom implementation functional.

### Feasibility in MAUI

While possible, implementing this would essentially require rewriting parts of the Windows-specific backend for file picking in MAUI. Developers would need to:

- Use Windows-specific libraries like `Windows.Win32` (via CsWin32 or WindowsAppSDK interop).
- Provide a custom interface for Windows file pickers that exposes the functionality of `IFileDialog::SetFolder`.

### Practical Alternative

For developers seeking this functionality without rewriting MAUI's internals, a viable approach is:

1. Create a **platform-specific implementation** for Windows within a MAUI app using dependency services.
2. Leverage the `IFileDialog::SetFolder` API through a helper library like [Vanara.PInvoke](https://github.com/dahall/Vanara) to simplify the interop.

**Comment by MartyIX** [comment]
Quick update for Windows:

* https://github.com/microsoft/WindowsAppSDK/pull/5155
* https://github.com/microsoft/WindowsAppSDK/pull/5547 (my understanding is that 1.8-exp3 should contain this feature)

**Comment by MartyIX** [comment]
Status update: https://github.com/microsoft/WindowsAppSDK/issues/88#issuecomment-3211470192 

> Looks like this is 1.8-preview1: https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/preview-channel#new-apis

Also, there is #30665. So it appears that default directory could be implemented for iOS + Windows per https://github.com/dotnet/maui/issues/9212#issuecomment-1826125726 for .NET MAUI v10 (if there is still time).

Anyone, who would attempt to do it, would achieve eternal glory. 🎉 

cc @bhavanesh2001 @kubaflo 👀 😆
